File: source/SharpFlame/ctrlPlayerNum.cs
using System;
using System.Windows.Forms;
using SharpFlame.FileIO;

namespace SharpFlame
{
    public partial class ctrlPlayerNum
    {
        public ToolStripButton[] tsbNumber = new ToolStripButton[11];

        private clsMap.clsUnitGroupContainer _Target;

        public const int ScavButtonNum = 10;

        public ctrlPlayerNum()
        {
            InitializeComponent();

            int A = 0;
            int B = 0;
            int ButtonsPerRow = 5;

            for ( A = 0; A <= ButtonsPerRow - 1; A++ )
            {
                tsbNumber[A] = new ToolStripButton();
                tsbNumber[A].DisplayStyle = ToolStripItemDisplayStyle.Text;
                tsbNumber[A].Text = IOUtil.InvariantToString_int(A);
                tsbNumber[A].AutoToolTip = false;
                tsbNumber[A].Click += tsbNumber_Clicked;
                tsPlayerNum1.Items.Add(tsbNumber[A]);

                B = A + ButtonsPerRow;
                tsbNumber[B] = new ToolStripButton();
                tsbNumber[B].DisplayStyle = ToolStripItemDisplayStyle.Text;
                tsbNumber[B].Text = IOUtil.InvariantToString_int(B);
                tsbNumber[B].AutoToolTip = false;
                tsbNumber[B].Click += tsbNumber_Clicked;
                tsPlayerNum2.Items.Add(tsbNumber[B]);
            }

            A = 10;

            tsbNumber[A] = new ToolStripButton();
            tsbNumber[A].DisplayStyle = ToolStripItemDisplayStyle.Text;
            tsbNumber[A].Text = "S";
            tsbNumber[A].AutoToolTip = false;
            tsbNumber[A].Click += tsbNumber_Clicked;
            tsPlayerNum2.Items.Add(tsbNumber[A]);

            Width = 24 * 6;
            Height = 25 * 2;
        }

        private void tsbNumber_Clicked(object sender, EventArgs e)
        {
            if ( _Target == null )
            {
                return;
            }

            ToolStripButton tsb = (ToolStripButton)sender;
            clsMap.clsUnitGroup UnitGroup = (clsMap.clsUnitGroup)tsb.Tag;

            _Target.Item = UnitGroup;
        }

        public clsMap.clsUnitGroupContainer Target
        {
            get { return _Target; }
            set
            {
                if ( value == _Target )
                {
                    return;
                }
                if ( _Target != null )
                {
                    _Target.Changed -= SelectedChanged;
                }
                _Target = value;
                if ( _Target != null )
                {
                    _Target.Changed += SelectedChanged;
                }
                SelectedChanged();
            }
        }

        private void SelectedChanged()
        {
            int A = 0;
            clsMap.clsUnitGroup UnitGroup = default(clsMap.clsUnitGroup);

            if ( _Target == null )
            {
                UnitGroup = null;
            }
            else
            {
                UnitGroup = _Target.Item;
            }

            if ( UnitGroup == null )
            {
                for ( A = 0; A <= 10; A++ )
                {
                    tsbNumber[A].Checked = false;
                }
            }
            else
            {
                for ( A = 0; A <= 10; A++ )
                {
                    tsbNumber[A].Checked = ((clsMap.clsUnitGroup)(tsbNumber[A].Tag)) == UnitGroup;
                }
            }
        }

        public void SetMap(clsMap NewMap)
        {
            int A = 0;

            if ( NewMap == null )
            {
                for ( A = 0; A <= modProgram.PlayerCountMax - 1; A++ )
                {
                    tsbNumber[A].Tag = null;
                }
                tsbNumber[ScavButtonNum].Tag = null;
            }
            else
            {
                for ( A = 0; A <= modProgram.PlayerCountMax - 1; A++ )
                {
                    tsbNumber[A].Tag = NewMap.UnitGroups[A];
                }
                tsbNumber[ScavButtonNum].Tag = NewMap.ScavengerUnitGroup;
            }
            SelectedChanged();
        }
    }
}

File: reference/FlaMEVB/ctrlPlayerNum.vb

Public Class ctrlPlayerNum

    Public tsbNumber(10) As ToolStripButton

    Private _Target As clsMap.clsUnitGroupContainer

    Public Const ScavButtonNum As Integer = 10

    Public Sub New()
        InitializeComponent()

        Dim A As Integer
        Dim B As Integer
        Dim ButtonsPerRow As Integer = 5

        For A = 0 To ButtonsPerRow - 1
            tsbNumber(A) = New ToolStripButton
            tsbNumber(A).DisplayStyle = ToolStripItemDisplayStyle.Text
            tsbNumber(A).Text = InvariantToString_int(A)
            tsbNumber(A).AutoToolTip = False
            AddHandler tsbNumber(A).Click, AddressOf tsbNumber_Clicked
            tsPlayerNum1.Items.Add(tsbNumber(A))

            B = A + ButtonsPerRow
            tsbNumber(B) = New ToolStripButton
            tsbNumber(B).DisplayStyle = ToolStripItemDisplayStyle.Text
            tsbNumber(B).Text = InvariantToString_int(B)
            tsbNumber(B).AutoToolTip = False
            AddHandler tsbNumber(B).Click, AddressOf tsbNumber_Clicked
            tsPlayerNum2.Items.Add(tsbNumber(B))
        Next

        A = 10

        tsbNumber(A) = New ToolStripButton
        tsbNumber(A).DisplayStyle = ToolStripItemDisplayStyle.Text
        tsbNumber(A).Text = "S"
        tsbNumber(A).AutoToolTip = False
        AddHandler tsbNumber(A).Click, AddressOf tsbNumber_Clicked
        tsPlayerNum2.Items.Add(tsbNumber(A))

        Width = 24 * 6
        Height = 25 * 2
    End Sub

    Private Sub tsbNumber_Clicked(sender As Object, e As EventArgs)

        If _Target Is Nothing Then
            Exit Sub
        End If

        Dim tsb As ToolStripButton = CType(sender, ToolStripButton)
        Dim UnitGroup As clsMap.clsUnitGroup = CType(tsb.Tag, clsMap.clsUnitGroup)

        _Target.Item = UnitGroup
    End Sub

    Public Property Target As clsMap.clsUnitGroupContainer
        Get
            Return _Target
        End Get
        Set(value As clsMap.clsUnitGroupContainer)
            If value Is _Target Then
                Exit Property
            End If
            If _Target IsNot Nothing Then
                RemoveHandler _Target.Changed, AddressOf SelectedChanged
            End If
            _Target = value
            If _Target IsNot Nothing Then
                AddHandler _Target.Changed, AddressOf SelectedChanged
            End If
            SelectedChanged()
        End Set
    End Property

    Private Sub SelectedChanged()

        Dim A As Integer
        Dim UnitGroup As clsMap.clsUnitGroup

        If _Target Is Nothing Then
            UnitGroup = Nothing
        Else
            UnitGroup = _Target.Item
        End If

        If UnitGroup Is Nothing Then
            For A = 0 To 10
                tsbNumber(A).Checked = False
            Next
        Else
            For A = 0 To 10
                tsbNumber(A).Checked = (CType(tsbNumber(A).Tag, clsMap.clsUnitGroup) Is UnitGroup)
            Next
        End If
    End Sub

    Public Sub SetMap(NewMap As clsMap)
        Dim A As Integer

        If NewMap Is Nothing Then
            For A = 0 To PlayerCountMax - 1
                tsbNumber(A).Tag = Nothing
            Next
            tsbNumber(ScavButtonNum).Tag = Nothing
        Else
            For A = 0 To PlayerCountMax - 1
                tsbNumber(A).Tag = NewMap.UnitGroups.Item(A)
            Next
            tsbNumber(ScavButtonNum).Tag = NewMap.ScavengerUnitGroup
        End If
        SelectedChanged()
    End Sub
End Class
